src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java

```diff
@@ -0,0 +1,62 @@
+package com.project.portfolio.springboot.onlinesellingandshopping.serverapp.security;
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.AuthGroup;
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.User;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.*;
+
+public class ApplicationUserPrincipal implements UserDetails {
+
+    private final User user;
+    private final List<AuthGroup> authGroups;
+
+    public ApplicationUserPrincipal(User user, List<AuthGroup> authGroups) {
+        super();
+        this.user = user;
+        this.authGroups = authGroups;
+    }
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        if(null==authGroups){
+            return Collections.emptySet();
+        }
+        Set<SimpleGrantedAuthority> grantedAuthorities = new HashSet<>();
```

Comment:
it's better to add a 'final' modifier to the local variables to signal you don't intend re-assigning it later in the method block.